Add customizable shortcuts panel

ollimeier opened the pull request:
This fixes #1594, fixes #1589, fixes #721, and fixes #1642.

Known issues:
- `event.preventDefault()` does not work for default shortcuts like cmd+N (event.stopImmediatePropagation() does not work either) 
-> **some default shortcuts cannot be overwritten**, please see: [bugs.chromium](https://bugs.chromium.org/p/chromium/issues/detail?id=33056&no_tracker_redirect=1)
-> [Good comment related to this issue]( https://github.com/liftoff/GateOne/issues/290#issuecomment-23121578)

justvanrossum commented on src/fontra/client/lang/en.json:
```diff
@@ -60,6 +73,7 @@
   "editor.pointer-tool": "Pointer Tool",
   "editor.power-ruler-tool": "Power Ruler Tool",
+  "editor.shape-tool": "Shape Tool",
   "editor.shape-tool-ellipse": "Shape Tool Ellipsis",
```
This typo must have existed for a while...
```suggestion
  "editor.shape-tool-ellipse": "Shape Tool Ellipse",
```

justvanrossum commented on src/fontra/views/applicationsettings/panel-base.js:
```diff
@@ -0,0 +1,91 @@
+  initializePanel() {
+    this.undoStack = new UndoStack();
+
+    // const subscribePattern = Object.fromEntries(
```
These things should be deleted: it is when a view deals with a font. This view doesn't.

ollimeier replied:
I removed the commented out code.

justvanrossum reviewed (commented):
This isn't a proper review yet, I just noticed some small things.

justvanrossum commented on src/fontra/views/applicationsettings/panel-shortcuts.js:
```diff
@@ -0,0 +1,659 @@
+      }
+      shortCutsDataCustom[actionIdentifier] = shortCuts;
+    }
+    const data = JSON.stringify(shortCutsDataCustom);
```
Use indentation of 2 spaces:
```suggestion
    const data = JSON.stringify(shortCutsDataCustom, null, 2);
```

justvanrossum commented on src/fontra/views/applicationsettings/panel-shortcuts.js:
```diff
@@ -0,0 +1,659 @@
+    const url = URL.createObjectURL(blob);
+    const a = document.createElement("a");
+    a.href = url;
+    a.download = "shortcuts.json";
```
```suggestion
    a.download = "fontra-shortcuts.json";
```

justvanrossum reviewed (commented):
Two more small things.

justvanrossum commented:
I noticed that if you assign a shortcut to the ellipse tool, the tool will switch, but the icon doesn't. This is technically unrelated to this PR, but perhaps it stll makes sense to fix it here.

ollimeier commented:
> I noticed that if you assign a shortcut to the ellipse tool, the tool will switch, but the icon doesn't. This is technically unrelated to this PR, but perhaps it stll makes sense to fix it here.

I agree. I fixed it here already: https://github.com/googlefonts/fontra/pull/1604/files
It's related to `collapseSubTools`.

justvanrossum commented:
> I fixed it here already: https://github.com/googlefonts/fontra/pull/1604/files

Ah, but that PR got merged into a branch which is now abandoned, right?

ollimeier commented:
> > I fixed it here already: https://github.com/googlefonts/fontra/pull/1604/files
> 
> Ah, but that PR got merged into a branch which is now abandoned, right?

yes.

justvanrossum commented on src/fontra/client/core/actions.js:
```diff
@@ -187,6 +223,9 @@ function loadActionsByKeyOrCode() {
   actionsByKeyOrCode = {};
   for (const [actionIdentifier, action] of Object.entries(actionInfoController.model)) {
     for (const shortCut of action.customShortCuts || action.defaultShortCuts || []) {
+      if (!shortCut) {
```
How can this happen?

ollimeier replied:
That was related to customShortCuts being `null`. I changed it to be an empty list instead. The check `if (!shortCut)` is not needed anymore.

justvanrossum commented on src/fontra/client/core/utils.js:
```diff
@@ -120,14 +120,18 @@ export function hyphenatedToLabel(s) {
+
+// Therefore use window.navigator https://developer.mozilla.org/en-US/docs/Web/API/Window/navigator
+export const isMac =
+  typeof navigator !== "undefined" && navigator.userAgent.indexOf("Mac") != -1;
```
I don't think we support browsers that don't support `navigator`, so I think that first check is redundant.

ollimeier replied:
`npm test` fails without it: _Exception during run: ReferenceError: navigator is not defined_

justvanrossum replied:
Perfect, makes sense!

justvanrossum commented on src/fontra/views/applicationsettings/panel-shortcuts.js:
```diff
@@ -0,0 +1,682 @@
+    this.append(
+      html.createDomElement("icon-button", {
+        "class": "fontra-ui-shortcuts-panel-icon",
+        "src": "/tabler-icons/trash.svg",
```
I think we should use an "X" icon instead: https://tabler.io/icons/icon/x

ollimeier replied:
done.

justvanrossum commented on src/fontra/client/core/actions.js:
```diff
@@ -27,19 +28,22 @@ actionInfoController.addListener((event) => {
-  registerActionInfo(actionIdentifier, actionInfo);
+  registerActionInfo(actionIdentifier, actionInfo, sortIndex);
   registerActionCallbacks(actionIdentifier, callback, enabled);
+  sortIndex++;
```
I think the sortIndex logic should be done by registerActionInfo() itself. And it would not need the sortIndex argument.

justvanrossum replied:
Perhaps each topic should have its own sortIndex counter.

I assume the topic in the UI are currently also sorted based on the sortIndex?

ollimeier replied:
I moved the sortIndex logic out of registerAction into registerActionInfo.

ollimeier replied:
As discussed, I moved the sorting out of actions into the UI code (shortcuts panel) – including the discussed refinements to topic indices (00X0-action-topics.XXX) etc.

justvanrossum commented on src/fontra/client/core/actions.js:
```diff
@@ -27,19 +28,22 @@ actionInfoController.addListener((event) => {
   actionInfoController.model[actionIdentifier] = {
     ...actionInfo,
     customShortCuts: storedActionInfo.customShortCuts,
+    sortIndex: sortIndex,
```
I think the caller should (optionally) be able to set the sortIndex itself. I would set `actionInfo.sortIndex` only when `actionInfo.sortIndex` is undefined.

ollimeier replied:
I added it `sortIndex: actionInfo.sortIndex === undefined ? sortIndex : actionInfo.sortIndex,` and used it for "action.add-component" -> so that "Add Component" and "Decompose Component" are below each other.

justvanrossum reviewed (commented):
Some feedback, it's not a complete review.

This doesn't support multiple shortcuts for a single action yet, right?

justvanrossum commented:
I added a help func named `getBaseKeyFromKeyEvent(event)` to actions.js. We should use its return value to:
- match against what it currently called `keyOrCode`, but should perhaps then be called `baseKey`
- display the "base" key as part of a menu item shortcut string, and in the custom shortcut UI

ollimeier commented:
> I added a helper func named `getBaseKeyFromKeyEvent(event)` to actions.js. We should use its return value to:
> 
> * match against what it currently called `keyOrCode`, but should perhaps then be called `baseKey`
> * display the "base" key as part of a menu item shortcut string, and in the custom shortcut UI

Thanks a lot for the helper function. I used it in the shortcuts panel to record the correct `keyOrCode`. But I don't fully understand your two bullet points. I don't know what else I should do with it. 

For example:
`match against what it currently called `keyOrCode`, but should perhaps then be called `baseKey``
-> Do you want me to replace `keyOrCode` with `baseKey` everywhere in the code? Or do you have a specific line of code mind which can be improved with `getBaseKeyFromKeyEvent`?

`display the "base" key as part of a menu item shortcut string, and in the custom shortcut UI`
-> In my opinion, this is already happening, isn't it? The keyOrCode (which should be the same as the baseKey) is used to get the correct icon/symbol via shortCutKeyMap.

I apologize that I am having trouble understanding these instructions.

ollimeier commented:
> I added a helper func named `getBaseKeyFromKeyEvent(event)` to actions.js. We should use its return value to:
> 
> * match against what it currently called `keyOrCode`, but should perhaps then be called `baseKey`
> * display the "base" key as part of a menu item shortcut string, and in the custom shortcut UI



> I added a helper func named `getBaseKeyFromKeyEvent(event)` to actions.js. We should use its return value to:
> 
> * match against what it currently called `keyOrCode`, but should perhaps then be called `baseKey`
> * display the "base" key as part of a menu item shortcut string, and in the custom shortcut UI

Done.

ollimeier commented:
> Some feedback, it's not a complete review.
> 
> This doesn't support multiple shortcuts for a single action yet, right?



> Some feedback, it's not a complete review.
> 
> This doesn't support multiple shortcuts for a single action yet, right?

There is no UI for adding multiple shortcuts to a single action, BUT you can add multiple shortcuts to a json file and import it. It will then support multiple shortcuts (but it will only display the first one). 

Is this enough for now or do you want me to add a UI for being able to add and visualize multiple shortcuts? Or is this better for a follow up?

ollimeier commented:
> I noticed that if you assign a shortcut to the ellipse tool, the tool will switch, but the icon doesn't. This is technically unrelated to this PR, but perhaps it stll makes sense to fix it here.



> I noticed that if you assign a shortcut to the ellipse tool, the tool will switch, but the icon doesn't. This is technically unrelated to this PR, but perhaps it stll makes sense to fix it here.



> I noticed that if you assign a shortcut to the ellipse tool, the tool will switch, but the icon doesn't. This is technically unrelated to this PR, but perhaps it stll makes sense to fix it here.

done.

justvanrossum commented on src/fontra/client/core/actions.js:
```diff
@@ -61,55 +73,94 @@ export function getActionInfo(actionIdentifier) {
+// add 0-9 keys
+for (let i = 0; i <= 9; i++) {
+  shortCutKeyMap[`Digit${i}`] = `${i}`;
+  shortCutKeyMap[`Numpad${i}`] = `${i}`;
```
I think we should do something like this:
```suggestion
  shortCutKeyMap[`Numpad${i}`] = `Numpad-${i}`;
```
The mechanism distinquishes between "regular" digits and "numpad" digits, and I think we should show that difference.

I would also like to discuss the other Numpad codes/char mapping, such as NumpadMultiply.

justvanrossum reviewed (commented):
I can't immediately find where the "record" button width is defined, but we need more space...
<img width="319" alt="image" src="https://github.com/user-attachments/assets/3e78a785-6a0c-4c7b-965e-aa805bcb0be6">

Also, it probably shouldn't overflow like that.

I would also like to discuss the "to blur" or "not to blur" after we've recorded a shortcut. While the field currently does not blur (stays in focus) the contents stay gray, which is confusing.

justvanrossum commented:
We have layout issue with long labels:

<img width="375" alt="image" src="https://github.com/user-attachments/assets/f945d20a-bf9c-4bd0-a818-91f939c929fc">

I don't think we can use the content size, as this will break alignment across sections. But:
- The label could wrap
- We could make the label column wider

(I think we should do both)

justvanrossum commented on src/fontra/views/applicationsettings/applicationsettings.js:
```diff
@@ -0,0 +1,116 @@
+import { message } from "/web-components/modal-dialog.js";
+
+export class ApplicationSettingsController {
+  static async fromWebSocket() {
```
This method name makes no sense, as there is no websocket involved. 

Perhaps the html script should call `.start()`, and the `start()` method should do `await ensureLanguageHasLoaded`.

ollimeier replied:
Thanks. done.

justvanrossum commented on src/fontra/views/applicationsettings/applicationsettings.js:
```diff
@@ -0,0 +1,116 @@
+
+export class ApplicationSettingsController {
+  static async fromWebSocket() {
+    document.title = `Fontra Application Settings`;
```
The title is set in the HTML, and isn't dynamic. Is this line needed?

justvanrossum commented on src/fontra/views/applicationsettings/applicationsettings.js:
```diff
@@ -0,0 +1,116 @@
+    panel?.handleKeyDown?.(event);
+  }
+
+  async messageFromServer(headline, msg) {
```
This method can be deleted: we have no live websocket connection to the server.

ollimeier replied:
done.

justvanrossum commented on src/fontra/views/applicationsettings/applicationsettings.js:
```diff
@@ -0,0 +1,116 @@
+    message(headline, msg);
+  }
+
+  handleRemoteClose(event) {
```
Same.

ollimeier replied:
done.

justvanrossum commented on src/fontra/views/applicationsettings/applicationsettings.js:
```diff
@@ -0,0 +1,116 @@
+    //
+  }
+
+  handleRemoteError(event) {
```
Same.

ollimeier replied:
done.

justvanrossum commented on src/fontra/views/applicationsettings/panel-base.js:
```diff
@@ -0,0 +1,57 @@
+    }
+  }
+
+  async doUndoRedo(isRedo) {
```
Is undo actually hooked up?

ollimeier replied:
It's not hooked up. It was pretty much a copy of the fontinfo panel-base.js. I left it almost unchanged, because I was not sure if the plan is to implement such functionality or not. 
I removed it.

justvanrossum commented on src/fontra/views/applicationsettings/panel-base.js:
```diff
@@ -0,0 +1,57 @@
+    this.setupUI();
+  }
+
+  async postChange(change, rollbackChange, undoLabel) {
```
I don't think this is used.

ollimeier replied:
done.

ollimeier replied:
I removed it.

ollimeier commented:
> We have a layout issue with long labels:
> 
> <img alt="image" width="375" src="https://private-user-images.githubusercontent.com/4246121/367881608-f945d20a-bf9c-4bd0-a818-91f939c929fc.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjY1NTU3MjQsIm5iZiI6MTcyNjU1NTQyNCwicGF0aCI6Ii80MjQ2MTIxLzM2Nzg4MTYwOC1mOTQ1ZDIwYS1iZjljLTRiZDAtYTgxOC05MWY5MzljOTI5ZmMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDkxNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA5MTdUMDY0MzQ0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YjZhYWM0YjFlYjFhZGE5MjU5MzBhMzhjMzZlZDJhNjE0MzMxOWNjYmZkODg3YjNlYWY3MWViYjVmNDg5YzQ2YSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.n8QO49a3r7QDbkLYODMe0lG6M32O_Yuso9_4OxfvJ78">
> I don't think we can use the content size, as this will break alignment across sections. But:
> 
> * The label could wrap
> * We could make the label column wider
> 
> (I think we should do both)

Thanks for the feedback. done.

justvanrossum commented on src/fontra/views/applicationsettings/applicationsettings.html:
```diff
@@ -0,0 +1,75 @@
+    import { ApplicationSettingsController } from "/applicationsettings/applicationsettings.js";
+
+    async function startApp() {
+      window.applicationSettingsController =
```
`start()` doesn't return anything, so this statement should be done in two parts:
- assign `applicationSettingsController`
- call `start()` on `window.applicationSettingsController()`

ollimeier replied:
I changed it, but I am not 100% sure if it's right – why I keep this feedback open.

justvanrossum replied:
You got the await wrong. The constructor is (by definition) not async, but `start()` is (as per `async` keyword.

ollimeier replied:
sorry and thanks. done.

justvanrossum commented on src/fontra/views/applicationsettings/panel-shortcuts.js:
```diff
@@ -0,0 +1,462 @@
+`);
+
+export class ShortCutsPanel extends BaseInfoPanel {
+  static title = "shortcuts.title";
```
I know we're not (yet) doing this in the fontinfo view, but let's use a prefix for all translatable string keys for this view.

```suggestion
  static title = "application-settings.shortcuts.title";
```

(This obviously also requires changes to the language files)

justvanrossum commented on src/fontra/views/applicationsettings/panel-shortcuts.js:
```diff
@@ -0,0 +1,462 @@
+      reader.onload = async (event) => {
+        const data = JSON.parse(event.target.result);
+        for (const actionIdentifier in data) {
+          setCustomShortCuts(actionIdentifier, data[actionIdentifier]);
```
Please add a check here to ensure `actionIdentifier` is an existing one. Else `setCustomShortCuts()` will fail.

ollimeier replied:
Didn't you say we won't check that because it's covered in `setCustomShortCuts` anyway (and good to know if there is something wrong):
`  assert(actionInfo, `unknown actionIdentifier: ${actionIdentifier}`);`
Please don't get me wrong, I am happy to change it. I only want to double check with you if I misunderstood something.

justvanrossum replied:
No. In the current state import will fail completely if it uses a non-existent key. This would happen if we'd rename or delete an action. So, the assert in `setCustomShortCuts()` is right, but the _caller_ can (and should in this case) filter for existing action identifiers.

ollimeier replied:
Thanks. done.

justvanrossum commented on src/fontra/views/applicationsettings/panel-shortcuts.js:
```diff
@@ -0,0 +1,462 @@
+    }
+  }
+
+  recordShortCutKeyup(id, event) {
```
```suggestion
  recordShortCutKeyUp(id, event) {
```

ollimeier replied:
done.

ollimeier commented on src/fontra/views/applicationsettings/panel-shortcuts.js:
```diff
@@ -0,0 +1,462 @@
+    element.focus();
+  }
+
+  outOfFocus(id) {
```
done.

ollimeier commented on src/fontra/views/editor/editor.js:
```diff
@@ -1186,7 +1220,15 @@ export class EditorController {
   }
 
-  setSelectedTool(toolIdentifier) {
+  collapseSubTools(editToolsElement) {
```
ok. done.

justvanrossum commented on src/fontra/client/lang/en.json:
```diff
@@ -99,6 +115,7 @@
   "menubar.view.select-next-source": "Select next source",
   "menubar.view.select-previous-source": "Select previous source",
   "selection.none": "(No selection)",
+  "server-info.title": "Server Info",
```
This one should also get the "application-settings" prefix.
